src/target/riscv/riscv.c

```diff
@@ -598,90 +606,174 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
+								  struct trigger_request_info trig2_info)
+{
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	RISCV_INFO(r);
+
+	for (unsigned int idx = 0;
+		 find_next_trigger(target, trigger_type, true, &idx) == ERROR_OK &&
+		 ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+		 ++idx) {
+		ret = set_trigger(target, idx, trig1_info.tdata1, trig1_info.tdata2,
+						  trig1_info.tdata1_ignore_mask);
+		if (ret != ERROR_OK)
+			continue;
+		ret = set_trigger(target, idx + 1, trig2_info.tdata1, trig2_info.tdata2,
+						  trig2_info.tdata1_ignore_mask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			r->trigger_unique_id[idx + 1] = trigger->unique_id;
+			return ERROR_OK;
+		} else
+			set_trigger(target, idx, 0, 0,
+						0); /* Undo the setting of the previous trigger*/
 	}
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_LT);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	tdata2 = trigger->address + trigger->length;
-	ret = set_trigger(target, idx + 1, tdata1, tdata2, 0);
-	if (ret != ERROR_OK) {
-		/* Undo the setting of the previous trigger*/
-		set_trigger(target, idx, 0, 0, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
-		return ret;
+	return ret;
+}
+
+static int try_trigger_request_match_with_fallbacks(
+	struct target *target, struct trigger *trigger, int trigger_type,
+	struct trigger_request_info napot_trigger_info,
+	struct trigger_request_info trig1_ge_info,
+	struct trigger_request_info trig2_lt_info,
+	struct trigger_request_info trig1_lt_info,
+	struct trigger_request_info trig2_ge_info,
+	struct trigger_request_info trig_eq_info)
+{
+	int ret = ERROR_OK;
+
+	if (!(trigger->execute || trigger->length == 1)) {
+		if (can_use_napot_match(trigger)) {
+			napot_trigger_info.tdata2 =
+				trigger->address | ((trigger->length - 1) >> 1);
+			LOG_TARGET_DEBUG(target, "trying t request NAPOT match");
+			ret = try_trigger_request_match(target, trigger, trigger_type,
+											napot_trigger_info);
+			if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+				return ret;
+		}
+		LOG_TARGET_DEBUG(target, "trying t request GE/LT match");
+		ret = try_chained_trigger_request_match(target, trigger, trigger_type,
+												trig1_ge_info, trig2_lt_info);
+		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			return ret;
+
+		LOG_TARGET_DEBUG(target, "trying t request LT/GE match");
+		ret = try_chained_trigger_request_match(target, trigger, trigger_type,
+												trig1_lt_info, trig2_ge_info);
+		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			return ret;
 	}
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	r->trigger_unique_id[idx + 1] = trigger->unique_id;
-	return ERROR_OK;
+	LOG_TARGET_DEBUG(target, "trying t request EQUAL match");
+	return try_trigger_request_match(target, trigger, trigger_type,
+									 trig_eq_info);
+}
```

Comment:
I recommend to remove this function completely and put its logic directly to `maybe_add_trigger_t2` and `maybe_add_trigger_t6`. The reason is that now all the possible trigger combinations are always created and passed to this function, even if they do not make sense for the given scenario.

For instance, if we have an execute=1 trigger, all the `napot_trigger_info`, `trig1_ge_info` ... structures are created, passed to this function but are irrelevant - and that decision is made only at the very end of the chain. This makes it harder to whoever reads the code to understand it properly.

Once this function is deleted, the `maybe_add_trigger_tX()` function will look like this (pseudocode):

```c
int maybe_add_trigger_tX( /*...*/ )
{
    /* Prepare the common trigger structure */
    /* ... */

    /* Prepare the exact equality match trigger */
    /* trig_eq_info */

    if (execute==1 || size == 1) {
        /* HW watchpoint of size=1 or HW breakpoint. Use simply an exact-match trigger. */
        return try_setup_single_match_trigger( /*...*/ );
    }

    /* Otherwise this is a watchpoint of size > 1. 
     * Try multiple trigger variants to cover this case. */

    /* Find out if we can use NAPOT match */
    if (can_use_napot_match( /* ... */ )) {
        /* Prepare napot_trigger_info */
        /* ... */

        ret = try_setup_single_match_trigger( /* ... */ );
        if (ret == ERROR_OK)
            return ERROR_OK;  /* Done - success */
    }

    /* NAPOT not possible or not supported by the target */

    /* Try a pair of chained triggers GE + LT */
    /* Prepare trig1_ge_info */
    /* ... */
    /* Prepare trig2_lt_info */
    /* ... */
    ret = try_setup_chained_match_triggers( /*...*/ );
    if (ret == ERROR_OK)
        return ERROR_OK;

    /* Try a pair of chained triggers LT + GE */
    /* Prepare trig1_lt_info */
    /* ... */
    /* Prepare trig2_ge_info */
    /* ... */
    ret = try_setup_chained_match_triggers( /*...*/ );
    if (ret == ERROR_OK)
        return ERROR_OK; 

    /* Last fallback: Try at least an exact match against the beginning of the range */
    return try_setup_single_match_trigger( trigger_eq_info );
}
```

Reply:
Thank you for the suggestion.

 I would like to preserve the function in question. In my opinion, it is preferable to minimize code repetition, especially if the code has non-trivial control flow. Moving control flow into a separate function simplifies debugging, modification, and makes it clear what exactly is the difference between _t2 and _t6 variants.

Reply:
I'm not sure if I like this, but an alternate option is to make a struct that contains all the relevant bitmasks for MCONTROL and MCONTROL6, then pass that struct to a single function that performs all the logic.

Reply:
> I would like to preserve the function in question. In my opinion, it is preferable to minimize code repetition, especially if the code has non-trivial control flow. Moving control flow into a separate function simplifies debugging, modification, and makes it clear what exactly is the difference between _t2 and _t6 variants.

Thank you for your reply. I agree with your arguments but I'd like that to be achieved by a different code structure. My updated proposal, that takes into account your points, would be:

- Have a single `maybe_add_trigger_t2_t6()` function that has the structure shown above - this would contain the control flow and would cover triggers of both type=2 and type=6
- Have a helper function(s) that fill the data structures (`napot_trigger_info`, `trig1_ge_info` ...) depending on the trigger type but do not contain any control flow. This would be called from `maybe_add_trigger_t2_t6()`.

This will avoid code repetition, separate control flow and data, and make the code easier to grasp by showing the control flow first ("what happens and in what sequence") and filling the data structures will be secondary ("how that happens internally").

Reply:
Addressed. I've reconfigured the code in such a way, that all the modifications to `trigger_request_info`, necessary for `maybe_add_trigger_t2_t6` are represented in `match_triggers_field_info`.

Reply:
Thank you for the changes. I have submitted one last round of suggestions how to simplify the code even further.